### PR TITLE
BLD: Add windows-latest to GHA CI

### PR DESCRIPTION
**Pull request by gerrymanoim**
Closes https://github.com/ibis-project/ibis/issues/2389

This uses https://github.com/conda-incubator/setup-miniconda to bootstrap the conda environment for ubuntu and windows. 

Caveats and notes:

- I added `basic_setup_env.sh` because I wasn't sure if this was an acceptable solution. If it is, I can get rid of `basic_setup_env.sh`, make all my changes in `setup_env.sh` and update all the steps with the github action. 
- Right now windows requires you to `conda activate base` before running the script (otherwise `conda` is not found). I'm not sure why. 
- I tried configuring `setup-miniconda` to install the `environment.yml` directly (using the `environment-file` parameter) and to automatically activate `ibis-dev` using `activate-environment`. I could not get this to work on windows. 
- I originally tried to solve this by setting the correct `CONDA_PATH` inside `ci/setup_env.sh` but ran into issues with other things being/not being in the path and could not get the environment to cleanly install. If you look at the commits in https://github.com/ibis-project/ibis/pull/2407/commits (ignoring the last commit) you can see a few things I tried. 
- A cleaner solution might be to make powershell equivalents of `run_tests` and `setup_env` and run those when testing windows.

### CODE REVIEW

**Comment by datapythonista** [comment]
Thanks @gerrymanoim, sorry for the delay. Great job.

The solution looks good, I experimented with the conda action, but I found it easier to use conda directly, when implementing this for Linux only.

Can you get your changes merged to `setup_env.sh` as you say, so we don't have duplicate scripts. Thanks!

**Comment by gerrymanoim** [comment]
There's something between git-bash, conda, and windows that just wasn't working well. 

Yep - will make that change.

**Comment by gerrymanoim** [comment]
I fixed an issue in how I was passing in the python version, so I was able to clean up a lot and use the github action a bit better. 

I'm still not entirely clear on why windows requires you to use conda activate before running the script, perhaps something special with git-bash maybe not loading the bash profile correctly? 

Impala doesn't start, but that's covered by https://github.com/ibis-project/ibis/issues/2388. 

I don't believe that the azure failures are from anything I did, but let me know if that isn't the case.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -15,19 +15,28 @@ jobs:
...
         python_version: ["3.7", "3.8"]
     steps:
     - name: checkout
       uses: actions/checkout@v1
-
+    - uses: conda-incubator/setup-miniconda@v1
+      with:
+        auto-update-conda: true
+        python-version: ${{ matrix.python_version }}
+        activate-environment: "ibis-dev"
```
Comment (datapythonista) [inline .github/workflows/main.yml:27]: Can we use the base environment, so we don't need to activate in every step?

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -15,19 +15,28 @@ jobs:
...
+        environment-file: "environment.yml"
     - name: set up environment
-      run: ./ci/setup_env.sh "${{ matrix.python_version }}" "$BACKENDS"
-      shell: bash
+      run: |
+        conda activate ibis-dev
+        ./ci/setup_env.sh "${{ matrix.python_version }}" "$BACKENDS"
+      shell: bash -l {0}
```
Comment (datapythonista) [inline .github/workflows/main.yml:33]: Why is this required?

File: .github/workflows/main.yml (hunk 1 above)
Comment (gerrymanoim) [reply .github/workflows/main.yml:27]: For some reason, even with the base env I needed to activate the env in windows/git bash.

File: .github/workflows/main.yml (hunk 2 above)
Comment (gerrymanoim) [reply .github/workflows/main.yml:33]: via https://github.com/conda-incubator/setup-miniconda#important

> Bash shells do not use ~/.profile or ~/.bashrc so these shells need to be explicitely declared as shell: bash -l {0} on steps that need to be properly activated. This is because bash shells are executed with bash --noprofile --norc -eo pipefail {0} thus ignoring updated on bash profile files made by conda init bash. See Github Actions Documentation and thread.

**Comment by datapythonista** [comment]
Thanks for the information @gerrymanoim, that's very useful to know.

Do you remember what were the problems you had when trying to make `setup_env.sh` in Windows? This approach is introducing 80 lines in our CI file, which has around 250. The original approach of making `setup_env.sh` work on Windows should require minimal changes, so I'd like to explore that option further and make sure we can't get it working, before having to add the `bash` and `conda acitvate` lines to all our steps.

**Comment by gerrymanoim** [comment]
Makes sense - I really should have taken better notes while I was doing this, but didn't expect it to end up this complicated. 

Let me dig up why I originally went down this path.

**Comment by datapythonista** [comment]
This was finally implemented in #2500. Thanks for the work on this @gerrymanoim
